### PR TITLE
feat: support multiple optimizers and LR schedulers in Pytorch trial interface [DET-3195, DET-3196]

### PR DESCRIPTION
**Pull request by shiyuann**
## Description

This PR does the following four things.

1. support the Pytorch trial interface with multiple optimizers and LR schedulers. It adds two new functions `build_optimizers` and `build_lr_schedulers`. It keeps the old interface for backward compatibility. These functions will be started with an underscore for now until officially released. Users can return the only instance or a list of instances of `torch.optim.Optimizer` in `build_optimizers` and return the only instance or a list of instances of `determined.torch.LRScheduler`.

2. update the initialization for multiple optimizers and LR schedulers, including broadcasting the parameters of them and wrapping the optimizers with Horovod.

3. update the checkpoint serialization schema with backward compatibility.

4. update the Pytorch context interface.

## Test Plan

Use the test suite and CI.

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_trial.py [hunk 1]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
...
         check.is_not_none(self.device)
 
-    def _init_model_and_optimizer(self) -> None:
+    def _init_model_and_optimizers(self) -> None:
+        # backwards compatibility: remove this when removing the old functions.
+        if util.is_overridden(self.trial._build_optimizers, PyTorchTrial) or util.is_overridden(
+            self.trial._build_lr_schedulers, PyTorchTrial
+        ):
+            check.true(util.is_overridden(self.trial._build_optimizers, PyTorchTrial))
+            check.true(util.is_overridden(self.trial._build_lr_schedulers, PyTorchTrial))
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:131]: non-blocking: I think we should add checks that the old versions are not overridden also.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 2]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
             self.device = torch.device("cpu")
         check.is_not_none(self.device)
 
-    def _init_model_and_optimizer(self) -> None:
+    def _init_model_and_optimizers(self) -> None:
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:125]: blocking: function name should reflect that `lr_schedulers` are being configured here as well

File: harness/determined/pytorch/_pytorch_trial.py [hunk 3]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
...
+        else:
+            raise det.errors.InvalidExperimentException(
+                "invalid return type from build_optimizers. "
+                "Please see an instance list or the only instance of "
+                "torch.optim.Optimizer."
+            )
 
-        self._init_device()
-        self.context.model = self.context.model.to(self.device)
+        lr_schedulers = self.trial._build_lr_schedulers(optimizers)
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:155]: blocking: this ordering will lead to the schedulers pointing to the wrong optimizer in the cases where we use horovod. We need to move this after the horovod optimizer initialization.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 4]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
...
+        else:
+            raise det.errors.InvalidExperimentException(
+                "invalid return type from build_lr_schedulers. "
+                "Please see an instance list or the only instance of "
+                "determined.pytorch.LRScheduler."
+            )
+        # TODO: clean up _LRHelper
+        self.lr_helpers = [_LRHelper(lr_scheduler) for lr_scheduler in self.context.lr_schedulers]
+
+        def filter_named_parameters(
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:169]: non-blocking: would be useful to add a comment here.

**Comment by rb-determined-ai** [comment]
I added the user-facing API change tag, please remember to use that

File: harness/determined/pytorch/_pytorch_trial.py [hunk 5]
```diff
@@ -695,12 +754,12 @@ def _save(self, path: pathlib.Path) -> workload.Response:
         # optimizer.
         checkpoint = {
             "model_state_dict": self.context.model.state_dict(),
-            "optimizer_state_dict": self.context.optimizer.state_dict(),
+            "optimizers_state_dict": [
+                optimizer.state_dict() for optimizer in self.context.optimizers
+            ],
+            "lr_scheduler": [lr_helper.state_dict() for lr_helper in self.lr_helpers],
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:760]: blocking: `lr_scheduler` -> `lr_schedulers`

File: harness/determined/pytorch/_pytorch_trial.py [hunk 6]
```diff
@@ -657,8 +702,22 @@ def _load(self) -> None:
...
+        if "optimizer_state_dict" in checkpoint:
+            check.not_in("optimizers_state_dict", checkpoint)
+            check.eq(len(self.context.optimizers), 1)
+            self.context.optimizers[0].load_state_dict(checkpoint["optimizer_state_dict"])
+        else:
+            for idx, optimizer in enumerate(self.context.optimizers):
+                optimizer.load_state_dict(checkpoint["optimizers_state_dict"][idx])
+
+        # backward compatibility: support old schema of checkpointing the only lr scheduler.
+        if isinstance(checkpoint["lr_scheduler"], dict):
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:716]: non-blocking: we should use the same format here and optimizer loading

**Comment by shiyuann** [comment]
@aaron276h @rb-determined-ai Update this PR to include the Pytorch context interface to pass the unit tests.

File: harness/determined/pytorch/_pytorch_context.py [hunk 7]
```diff
@@ -44,10 +43,28 @@ def get_optimizer(self) -> torch.optim.Optimizer:  # type: ignore
...
+                raise det.errors.InvalidExperimentException(
+                    "Must specify the index in get_optimizer "
+                    "if there are more than one optimizers."
+                )
+        return self.optimizers[idx]
 
-    def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+    def _get_optimizers(self) -> List[torch.optim.Optimizer]:  # type: ignore
+        """
+        Get the optimizer associated with the trial. This function should not be
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_context.py:58]: non-blocking: Get the `optimizer(s) ..`

File: harness/determined/pytorch/_pytorch_context.py [hunk 8]
```diff
@@ -57,4 +74,24 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
...
+            else:
+                raise det.errors.InvalidExperimentException(
+                    "Must specify the index in get_lr_scheduler "
+                    "if there are more than one LR schedulers."
+                )
+        return self.lr_schedulers[idx]
+
+    def _get_lr_schedulers(self) -> List[pytorch.LRScheduler]:
+        """
+        Get the schedulers associated with the trial, if one is defined. This
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_context.py:89]: non-blocking: `Get the schedulers(s) ...`

File: harness/determined/pytorch/_pytorch_context.py [hunk 9]
```diff
@@ -35,7 +34,7 @@ def get_model(self) -> torch.nn.Module:
         check.check_not_none(self.model)
         return cast(torch.nn.Module, self.model)
 
-    def get_optimizer(self) -> torch.optim.Optimizer:  # type: ignore
+    def get_optimizer(self, idx: Optional[int] = None) -> torch.optim.Optimizer:  # type: ignore
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_context.py:37]: non-blocking: until we make the new API public, I think expanding this arg might be somewhat confusing.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 3 above)
Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py:148]: When we use Horovod, it replaces the original optimizer with the Horovod optimizer. The `lr_scheduler` maintains a reference to the optimizer that is passed in to it in the constructor, updating its `lr`. So if we build the schedulers before we create the Horovod optimizer, they will be updating the `lr` of the wrong optimizer(s).

File: harness/determined/pytorch/_pytorch_trial.py [hunk 10]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
...
+                "determined.pytorch.LRScheduler."
+            )
+        # TODO: clean up _LRHelper
+        self.lr_helpers = [_LRHelper(lr_scheduler) for lr_scheduler in self.context.lr_schedulers]
+
+        def filter_named_parameters(
+            model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+        ) -> List:
+            """
+            filter_named_parameters filters the named parameters of a specified optimizer out
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:166]: praise: very descriptive comment

File: harness/determined/pytorch/_pytorch_trial.py [hunk 11]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
...
+                "invalid return type from build_lr_schedulers. "
+                "Please see a list or the only instance of "
+                "determined.pytorch.LRScheduler."
+            )
+        # TODO: clean up _LRHelper
+        self.lr_helpers = [_LRHelper(lr_scheduler) for lr_scheduler in self.context.lr_schedulers]
+
+        def filter_named_parameters(
+            model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+        ) -> List:
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:164]: question: do you think it would be worthwile to write a unit test for this function?

Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py:150]: IMO, this function is only used here and probably won't be used anywhere else in the future. It's also very transparent and straightforward in what it does. So don't think it's necessary.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 12]
```diff
@@ -122,23 +122,67 @@ def _init_device(self) -> None:
...
+            self.context.optimizers = [optimizers]
+        elif isinstance(optimizers, list):
+            check.gt_eq(len(optimizers), 1)
+            self.context.optimizers = optimizers
+        else:
+            raise det.errors.InvalidExperimentException(
+                "invalid return type from build_optimizers. "
+                "Please see an instance list or the only instance of "
+                "torch.optim.Optimizer."
+            )
```
[the remark this answers is not in the document]
Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py]: See https://github.com/determined-ai/determined/pull/641/files#diff-9595d835d00481dce482c51a24f0038dR140

File: harness/determined/pytorch/_pytorch_trial.py (hunk 11 above)
Comment (shiyuann) [reply harness/determined/pytorch/_pytorch_trial.py:150]: I have manually tested to make sure it works.

**Comment by shiyuann** [comment]
Close this PR because we have update on the interface.

**Pull request closed without merging**